File: server/invite.test.ts
import { describe, it, expect, vi, beforeEach, afterEach } from "vitest";
import { createInvite, isInviteValid, consumeInvite } from "./invite.js";

describe("invite", () => {
  beforeEach(() => {
    vi.useFakeTimers();
    vi.setSystemTime(0);
  });
  afterEach(() => {
    vi.useRealTimers();
  });

  it("is valid when fresh and unused", () => {
    const invite = createInvite();
    expect(isInviteValid(invite, invite.value)).toBe(true);
  });

  it("is invalid after consume", () => {
    const invite = createInvite();
    consumeInvite(invite);
    expect(isInviteValid(invite, invite.value)).toBe(false);
  });

  it("is invalid after 15 minutes", () => {
    const invite = createInvite();
    vi.setSystemTime(15 * 60 * 1000 + 1);
    expect(isInviteValid(invite, invite.value)).toBe(false);
  });

  it("is invalid for wrong input", () => {
    const invite = createInvite();
    expect(isInviteValid(invite, "wrong")).toBe(false);
  });

  it("is invalid when invite is null", () => {
    expect(isInviteValid(null, "anything")).toBe(false);
  });
});

File: server-go/internal/auth/invite_test.go
package auth

import (
	"testing"
	"time"
)

func TestInvite_ValidWhenFreshAndUnused(t *testing.T) {
	inv := CreateInvite()
	if !inv.Valid(inv.Value) {
		t.Error("expected valid for fresh unused invite")
	}
}

func TestInvite_InvalidAfterConsume(t *testing.T) {
	inv := CreateInvite()
	inv.Consume()
	if inv.Valid(inv.Value) {
		t.Error("expected invalid after consume")
	}
}

func TestInvite_InvalidAfterExpiry(t *testing.T) {
	inv := &InviteToken{
		Value:     "test",
		ExpiresAt: time.Now().Add(-time.Millisecond),
	}
	if inv.Valid("test") {
		t.Error("expected invalid after expiry")
	}
}

func TestInvite_InvalidForWrongInput(t *testing.T) {
	inv := CreateInvite()
	if inv.Valid("wrong") {
		t.Error("expected invalid for wrong input")
	}
}

func TestInvite_NilIsInvalid(t *testing.T) {
	var inv *InviteToken
	if inv.Valid("anything") {
		t.Error("expected nil invite to be invalid")
	}
}
